python/ray/data/read_api.py

```diff
@@ -338,10 +344,14 @@ def read_datasource(
         read_tasks, ray_remote_args=ray_remote_args, owned_by_consumer=False
     )
 
+    read_op = Read(reader, requested_parallelism, ray_remote_args, read_args)
+    logical_plan = LogicalPlan(read_op)
```

Comment:
Seems we'd generate the read tasks twice when using the new path, should we add a TODO to avoid this in the future?

Reply:
It seems to me that the entire `get_read_tasks` logic should ideally be deferred to physical execution time.

Reply:
Added `TODO` for now.

Reply:
Agreed that we need more metadata for optimization. I am actually thinking we can expose those metadata ((number of files, file sizes, column types, number of columns, partition columns, etc.)) from either `Reader` or `Datasource`, for logical optimization.

The `ReadTask` is the generated function to actually read the file, which looks to me that we can defer to execution time.

Reply:
@c21 For sure, we'll just need to unbundle the read task generation a bit!